main.go

```diff
@@ -0,0 +1,393 @@
+			Subject:         u.PrimaryEmail,
+		})
+		if err != nil {
+			glog.Errorf("Error creating CRM credentials for user %s,  %v", u.PrimaryEmail, err)
+			return []*cloudresourcemanager.Project{}, err
+		}
+		crmService, err = cloudresourcemanager.NewService(ctx, option.WithTokenSource(ts))
+		if err != nil {
+			glog.Errorf("Error creating CRM Service for  user %s,  %v", u.PrimaryEmail, err)
+			return []*cloudresourcemanager.Project{}, err
+		}
+	} else {
+		cred, err := google.CredentialsFromJSONWithParams(ctx, svcAccountJSONBytes, google.CredentialsParams{
+			Scopes:  []string{cloudresourcemanager.CloudPlatformReadOnlyScope},
+			Subject: u.PrimaryEmail,
+		})
+		if err != nil {
+			glog.Errorf("Error creating CRM credentials for  user %s,  %v", u.PrimaryEmail, err)
+			return []*cloudresourcemanager.Project{}, err
+		}
+		crmService, err = cloudresourcemanager.NewService(ctx, option.WithCredentials(cred))
+		if err != nil {
+			glog.Errorf("Error creating CRM Service for  user %s,  %v", u.PrimaryEmail, err)
+			return []*cloudresourcemanager.Project{}, err
+		}
+	}
+
+	projects := make([]*cloudresourcemanager.Project, 0)
+	// do not apply a list filter here; we need to audit all projects
+	//  https://pkg.go.dev/google.golang.org/api@v0.58.0/cloudresourcemanager/v1#ProjectsListCall.Filter
+	noFilter := ""
+
+	req := crmService.Projects.List().Filter(noFilter).PageSize(maxPageSize)
+	if err := req.Pages(ctx, func(page *cloudresourcemanager.ListProjectsResponse) error {
+		projects = append(projects, page.Projects...)
+		if err := limiter.Wait(ctx); err != nil {
+			glog.Errorf("Error in rate limiter for user %s %v", u.PrimaryEmail, err)
+			return err
+		}
+		if ctx.Err() != nil {
+			glog.Errorf("Error in rate limiter for user %s %v", u.PrimaryEmail, ctx.Err())
+			return ctx.Err()
+		}
+		return nil
+	}); err != nil {
+		glog.Errorf("Error iterating visible projects for user %s %v", u.PrimaryEmail, err)
+		return []*cloudresourcemanager.Project{}, err
+	}
+
+	for _, p := range projects {
+		glog.V(50).Infof("             User %s has Project visibility to %s", u.PrimaryEmail, p.ProjectId)
+		if _, ok := allProjects[p.ProjectId]; !ok {
+			glog.V(2).Infof("             User [%s] has external project visibility to [projects/%d](%s)", u.PrimaryEmail, p.ProjectNumber, p.ProjectId)
+		}
+	}
+	return projects, nil
+}
+
+func main() {
+
+	flag.Parse()
+	defer glog.Flush()
+	ctx := context.Background()
+
+	// configure API rate limits
+	limiter = rate.NewLimiter(rate.Limit(maxRequestsPerSecond), burst)
+	if *organization == "" || *subject == "" || *cx == "" {
+		glog.Error("--organization and --cx and --subject   must be specified")
+		return
+	}
+	if (*serviceAccountFile == "" && *impersonatedServiceAccount == "") || (*serviceAccountFile != "" && *impersonatedServiceAccount != "") {
+		glog.Error("either --serviceAccountFile or --impersonatedServiceAccount must be specified")
+		return
+	}
+
+	if *serviceAccountFile != "" {
+		var err error
```

Comment:
Remove this, use assignment from function:

```go
svcAccountJSONBytes, err := ...
```

Reply:
I'm talking about the `var err error` line here -- it's not clear to me why you are allocating it before hand. You can remove this, and make line 242 `:=` instead.

Reply:
oh, ok. will fix in net push

Reply:
hm. maybe i can't,

i've already defined `svcAccountJSONBytes`  as the global var, i can't use a := here  in that line

```golang
	if *serviceAccountFile != "" {
		var err error
		svcAccountJSONBytes, err = ioutil.ReadFile(*serviceAccountFile)
		if err != nil {
			glog.Fatal(err)
		}
	}
```

i could

```
	if *serviceAccountFile != "" {
		bt, err := ioutil.ReadFile(*serviceAccountFile)
		if err != nil {
			glog.Fatal(err)
		}
      svcAccountJSONBytes = bt
	}
```
but i don't thnk that buys much

Reply:
Nah, it doesn't. Okay, just leave it. Not ideal, but since this is more "script" like rather than a real program, slightly less worried about globals.